File: python/leetcode/editor/en/20.valid-parentheses.py
# Given a string s containing just the characters '(', ')', '{', '}', '[' and ']
# ', determine if the input string is valid. 
# 
#  An input string is valid if: 
# 
#  
#  Open brackets must be closed by the same type of brackets. 
#  Open brackets must be closed in the correct order. 
#  Every close bracket has a corresponding open bracket of the same type. 
#  
# 
#  
#  Example 1: 
# 
#  
# Input: s = "()"
# Output: true
#  
# 
#  Example 2: 
# 
#  
# Input: s = "()[]{}"
# Output: true
#  
# 
#  Example 3: 
# 
#  
# Input: s = "(]"
# Output: false
#  
# 
#  
#  Constraints: 
# 
#  
#  1 <= s.length <= 10⁴ 
#  s consists of parentheses only '()[]{}'. 
#  
# 
#  👍 23595 👎 1688


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def isValid(self, s: str) -> bool:
        str_list = list(s)
        stack = []
        for sl in str_list:
            if sl in ['(', '[', '{']:
                stack.append(sl)
            else:
                if len(stack) == 0:
                    return False
                sp = stack.pop()
                if not ((sp == '(' and sl == ')') or (sp == '[' and sl == ']') or (sp == '{' and sl == '}')):
                    return False

        return True if len(stack) == 0 else False


s = Solution()

print(s.isValid(s="()"))
print(s.isValid(s="()[]{}"))
print(s.isValid(s="(]"))
# leetcode submit region end(Prohibit modification and deletion)

File: go/leetcode/20.Valid-Parentheses.go
package leetcode

// かっこがちゃんと閉じられてるのか判定する
// () = true
// (} = false
// ({[]}) = true

// stack で入れてくだけだと最終的にチェック判定をしないといけない。
// 左括弧だけを stack にいれてペア判定する。最終的に stack が空であるかを見ればいい。
func isValid(s string) bool {
	// stack := linkedlitstack.New()
	// closeToOpen := map[rune]rune{')': '(', ']': '[', '}': '{'}
	//
	// for _, c := range s {
	// 	if open, exists := closeToOpen[c]; exists {
	// 		if !stack.Empty() {
	// 			top, ok := stack.Pop()
	// 			if ok && top.(rune) != open {
	// 				return false
	// 			}
	// 		} else {
	// 			return false
	// 		}
	// 	} else {
	// 		stack.Push(c)
	// 	}
	// }
	//
	// return stack.Empty()

	return true
}
